runtime/doc/lua.txt

```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
     See also: ~
       • # https://github.com/npm/node-semver#ranges
 
+
+==============================================================================
+Lua module: iter                                                    *lua-iter*
+
+Iter:all({self}, {pred})                                          *Iter:all()*
+    Return true if all of the items in the iterator match the given predicate.
+
+    Parameters: ~
+      • {pred}  function(...):bool Predicate function. Takes all values
+                returned from the previous stage in the pipeline as arguments
+                and returns true if the predicate matches.
+
+Iter:any({self}, {pred})                                          *Iter:any()*
+    Return true if any of the items in the iterator match the given predicate.
+
+    Parameters: ~
+      • {pred}  function(...):bool Predicate function. Takes all values
+                returned from the previous stage in the pipeline as arguments
+                and returns true if the predicate matches.
+
+Iter:collect({self}, {opts})                                  *Iter:collect()*
```

Comment:
> I borrowed the collect terminology from Rust, and I quite like it, but I'm not sure if that's a standard term.

It's used in java too but it takes a `Collector`, which is sort of a `Reducer`: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html#collect(java.util.stream.Collector)

If it's restricted to return a table I'd tend to go with `:totable`, same as https://luafun.github.io/reducing.html#fun.totable, but for stdlib I think something that is more generally useful would be warranted

> It may not hurt to leave it in though. I like that you can implement collect using fold, but imo it's good to have a dedicated collect function that converts an iterator into a table (it's not immediately obvious that fold would do that).

Could also still call it collect with an opts table that takes an initial value and an accumulator. They could default to an empty table and table.insert.

Reply:
Renamed to `totable` in 341e7fef0d754c6fcfcff2adfabcb0f444f9ae81. Restored `fold` in eaa15d437e47c97c82af3eed7bb8defc1d884e62.

Reply:
>If it's restricted to return a table I'd tend to go with :totable, same as https://luafun.github.io/reducing.html#fun.totable, but for stdlib I think something that is more generally useful would be warranted

Because tables are so central in Lua, and because conversion to a table is slightly different implementation wise for list-like tables, map-like tables, and function iterators (for both performance and semantics reasons), I think having a dedicated `totable` makes sense. For a more generic interface, we have `fold`.